RandomMCPlayer.java

```diff
@@ -168,7 +170,192 @@ private int simPlay(int depthLimit) {
 			return score;
 		}
 	}
+
+	private float getGridUtility() {
+		float gridUtility = 0;
+		for (int row = 0; row < SIZE; row++) {
+			gridUtility += getHandUtility(getHandByRow(row), true);
+		}
+		for (int col = 0; col < SIZE; col++) {
+			gridUtility += getHandUtility(getHandByColumn(col), false);
+		}
+		return gridUtility;
+	}
+
+	private Card[] getHandByRow(int row) {
+		Card[] hand = new Card[SIZE];
+		for (int col = 0; col < SIZE; col++) {
+			hand[col] = grid[row][col];
+		}
+		return hand;
+	}
+
+	private Card[] getHandByColumn(int col) {
+		Card[] hand = new Card[SIZE];
+		for (int row = 0; row < SIZE; row++) {
+			hand[row] = grid[row][col];
+		}
+		return hand;
+	}
+
+	private float getHandUtility(Card[] hand, boolean isRow) {
+		float utility = 0;
+		int numCards = 0;
+		for (int i = 0; i < SIZE ; i++) {
+			if (hand[i] != null) numCards++;
+		}
+		if (numCards == 0) utility = 0;
+		else if (numCards == 5) utility = system.getHandScore(hand);
+		else if (numCards == 4) utility = getExpectedValue(hand);
+		else utility = 3; // temp
+		return utility;
+	}
+
+	public float getExpectedValue(Card[] hand) {
+		float expectedValue = 0;
+		Integer [] suitArr  = new Integer[4];
+		Integer [] rankArr  = new Integer[13];
+
+		for (int i = 0; i < 5; i++) {
+			if (hand[i] != null) {
+				suitArr[hand[i].getSuit()]++;
+				rankArr[hand[i].getRank()]++;
+			}
+		}
+		ArrayList<Integer> suitArrayList = new ArrayList<>(Arrays.asList(suitArr));
+		ArrayList<Integer> rankArrayList = new ArrayList<>(Arrays.asList(rankArr));
+
+		if (suitArrayList.contains(4)){
+			expectedValue += probOfRoyalFlush(hand, suitArrayList.indexOf(4)) * 30; // royal flush
+			expectedValue += probOfSequence(hand, suitArrayList.indexOf(4)) * 30; // straight flush
+			expectedValue += probOfSuit(suitArrayList.indexOf(4)) * 5; // flush
+		}
+		
+		if (rankArrayList.contains(3)) {
+			expectedValue += probOfRank(rankArrayList.indexOf(3)) * 16; // fourOfKind
+			expectedValue += probOfRank(rankArrayList.indexOf(1)) * 10; // find card to complete 2 card rank for full-house 
+		}
+		else if (rankArrayList.contains(2)) {
+			if (Collections.frequency(rankArrayList, 2) == 2) { // check if there's two occurence of 2 
+				expectedValue += probOfRank(rankArrayList.indexOf(2), rankArrayList.lastIndexOf(2)) * 10; // find card for 3 card rank of full-house
+			}
+			else {
+				expectedValue += probOfRank(rankArrayList.lastIndexOf(2)) * 6; // threeOfKind
+				expectedValue += probOfRank(rankArrayList.lastIndexOf(1)) * 3; // twoPair
+			}
+		}
+		else {
+			expectedValue += probOfRank(rankArrayList) * 1; // onePair
+		}
+
+		expectedValue += probOfSequence(hand, -1) * 12; // straight flush
+
+		return expectedValue;
+	}
+
+	float probOfRoyalFlush(Card[] hand, int suit) {
+		int rankToGet = 0;
```

Comment:
For the royal flush check, you can hard-code a check on whether you have a ten, jack, queen, king, and ace and figure out which one is missing.